### PR TITLE
feat: Add AppInfoService (only the request and storage of data on device)

### PR DESCRIPTION
**Pull request by BiancaMihaila**
**Ticket Link:** 
[DCMAW-10332](https://govukverify.atlassian.net/browse/DCMAW-10332)

**Description Of Changes:**
- add AppInfoApi class: making the request to the appInfo endpoint
- create Remote and Local sources for AppInfo: create AppInfoRemoteSource for the requests and logic if this fails and create AppInfoLocalSource - fallback when the network request fails
- implementation of the AppInfoService: logic for how to handle remote and local sources

**Evidence:**
I cannot show the error being displayed when both local and remote fail - but it has been unit tested

https://github.com/user-attachments/assets/ffacd9f9-cc6b-44cd-87a3-732690eb3281

```
2024-10-10 15:54:57.118 10600-10600 AppInfoLocal            uk.gov.onelogin.build                D  AppInfoData(apps=App(android=AppInfo(minimumVersion=0.0.0, releaseFlags=ReleaseFlags(walletVisibleViaDeepLink=true, walletVisibleIfExists=true, walletVisibleToAll=false), available=true, featureFlags=FeatureFlags(appCheckEnabled=true))))
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  HttpClient: REQUEST: https://mobile.build.account.gov.uk/appInfo
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  METHOD: HttpMethod(value=GET)
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  COMMON HEADERS
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  -> Accept: application/json
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  -> Accept-Charset: UTF-8
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  -> Cache-Control: no-store
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  -> Content-Type: application/json
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  -> Strict-Transport-Security: max-age=31536000
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  -> User-Agent: OneLogin/v1.0.0 Google/Pixel 4a Android/33 uk.gov.onelogin.build/v1.0.0
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  -> X-Content-Type-Options: nosniff
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  -> X-Frame-Options: DENY
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  CONTENT HEADERS
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  -> Content-Length: 0
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  BODY Content-Type: null
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  BODY START
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  
2024-10-10 15:54:57.209 10600-10600 System.out              uk.gov.onelogin.build                I  BODY END
2024-10-10 15:54:57.214 10600-10600 Choreographer           uk.gov.onelogin.build                I  Skipped 63 frames!  The application may be doing too much work on its main thread.
2024-10-10 15:54:57.250 10600-10600 OnBackInvokedCallback   uk.gov.onelogin.build                W  OnBackInvokedCallback is not enabled for the application.
                                                                                                    Set 'android:enableOnBackInvokedCallback="true"' in the application manifest.
2024-10-10 15:54:57.292 10600-10650 TrafficStats            uk.gov.onelogin.build                D  tagSocket(107) with statsTag=0xffffffff, statsUid=-1
2024-10-10 15:54:57.975 10600-10651 System.out              uk.gov.onelogin.build                I  HttpClient: RESPONSE: 200 OK
2024-10-10 15:54:57.975 10600-10651 System.out              uk.gov.onelogin.build                I  METHOD: HttpMethod(value=GET)
2024-10-10 15:54:57.975 10600-10651 System.out              uk.gov.onelogin.build                I  FROM: https://mobile.build.account.gov.uk/appInfo
2024-10-10 15:54:57.975 10600-10651 System.out              uk.gov.onelogin.build                I  COMMON HEADERS
2024-10-10 15:54:57.975 10600-10651 System.out              uk.gov.onelogin.build                I  -> cache-control: no-store
2024-10-10 15:54:57.975 10600-10651 System.out              uk.gov.onelogin.build                I  -> connection: keep-alive
2024-10-10 15:54:57.975 10600-10651 System.out              uk.gov.onelogin.build                I  -> content-length: 609
2024-10-10 15:54:57.975 10600-10651 System.out              uk.gov.onelogin.build                I  -> content-type: application/json
2024-10-10 15:54:57.975 10600-10651 System.out              uk.gov.onelogin.build                I  -> date: Thu, 10 Oct 2024 14:54:57 GMT
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I  -> strict-transport-security: max-age=31536000
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I  -> x-amz-apigw-id: fcGeQEFWLPEEJ-g=
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I  -> x-amzn-requestid: c372ba74-426b-4b5e-a802-418e8cd5b2ef
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I  -> x-amzn-trace-id: Root=1-6707eac1-6d28fe9d08e713185fb1a137
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I  -> x-android-received-millis: 1728572097896
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I  -> x-android-response-source: NETWORK 200
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I  -> x-android-selected-protocol: http/1.1
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I  -> x-android-sent-millis: 1728572097738
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I  -> x-content-type-options: nosniff
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I  -> x-frame-options: DENY
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I  BODY Content-Type: application/json
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I  BODY START
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I  {
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I    "apps": {
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I      "android": {
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I        "minimumVersion": "0.0.0",
2024-10-10 15:54:57.976 10600-10651 System.out              uk.gov.onelogin.build                I        "releaseFlags": {
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I          "walletVisibleViaDeepLink": true,
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I          "walletVisibleIfExists": true,
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I          "walletVisibleToAll": false
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I        },
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I        "available": true,
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I        "featureFlags" : {
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I          "appCheckEnabled": true
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I        }
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I      },
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I      "iOS": {
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I        "minimumVersion": "1.0.0",
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I        "releaseFlags": {
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I          "walletVisibleViaDeepLink": true,
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I          "walletVisibleIfExists": true,
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I          "walletVisibleToAll": false
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I        },
2024-10-10 15:54:57.977 10600-10651 System.out              uk.gov.onelogin.build                I        "available": true,
2024-10-10 15:54:57.978 10600-10651 System.out              uk.gov.onelogin.build                I        "featureFlags" : {
2024-10-10 15:54:57.978 10600-10651 System.out              uk.gov.onelogin.build                I          "appCheckEnabled": true
2024-10-10 15:54:57.978 10600-10651 System.out              uk.gov.onelogin.build                I        }
2024-10-10 15:54:57.978 10600-10651 System.out              uk.gov.onelogin.build                I      }
2024-10-10 15:54:57.978 10600-10651 System.out              uk.gov.onelogin.build                I    }
2024-10-10 15:54:57.978 10600-10651 System.out              uk.gov.onelogin.build                I  }
2024-10-10 15:54:57.978 10600-10651 System.out              uk.gov.onelogin.build                I  
2024-10-10 15:54:57.978 10600-10651 System.out              uk.gov.onelogin.build                I  BODY END
2024-10-10 15:54:58.009 10600-10600 AppInfoRemote           uk.gov.onelogin.build                D  AppInfoData(apps=App(android=AppInfo(minimumVersion=0.0.0, releaseFlags=ReleaseFlags(walletVisibleViaDeepLink=true, walletVisibleIfExists=true, walletVisibleToAll=false), available=true, featureFlags=FeatureFlags(appCheckEnabled=true))))
```

https://github.com/user-attachments/assets/4a0e670e-2d93-4f49-a891-6ead01a2266a

```
2024-10-10 15:49:17.323  9927-9927  AppInfoLocal            uk.gov.onelogin.build                D  AppInfoData(apps=App(android=AppInfo(minimumVersion=0.0.0, releaseFlags=ReleaseFlags(walletVisibleViaDeepLink=true, walletVisibleIfExists=true, walletVisibleToAll=false), available=true, featureFlags=FeatureFlags(appCheckEnabled=true))))

```

https://github.com/user-attachments/assets/8ca8bad8-5159-409f-85d2-6063bf674853

**Definition Of Done Checklist:**

- [ ] Unit tests written to at least 80% code coverage (unless explicit approval given from TL that this is not necessary)
- [x] Relevant integration and end-to-end tests are written
- [x] Tests cover each acceptance criteria on the user story
- [ ] Demo completed by running code locally, demoing to a code reviewer & designer
- [ ] Sonar coverage gates met 
- [ ] If feature flagged, Feature is enabled in staging and manually tested to ensure integration

Resolves: DCMAW-10332

[DCMAW-10332]: https://govukverify.atlassian.net/browse/DCMAW-10332?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/src/build/res/values/login_flow.xml [hunk 1]
```diff
@@ -8,4 +8,5 @@
     <string name="webBaseHost" translatable="false">mobile.build.account.gov.uk</string>
     <string name="apiBaseUrl" translatable="false">https://mobile.build.account.gov.uk%1$s</string>
     <string name="helloWorldUrl" translatable="false">https://hello-world.token.build.account.gov.uk%1$s</string>
+    <string name="appInfoUrl" translatable="false">https://mobile.build.account.gov.uk%1$s</string>
```
Comment (alex-bradbury-gds) [inline app/src/build/res/values/login_flow.xml:11]: think you can just use the `apiBaseUrl` above

File: app/src/main/java/uk/gov/onelogin/appinfo/source/data/AppInfoLocalSourceImpl.kt [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            } else {
+                AppInfoLocalState.Failure(APP_INFO_LOCAL_SOURCE_ERROR)
+            }
+        } catch (e: ClassCastException) {
+            AppInfoLocalState.Failure(APP_INFO_CLASS_CAST_ERROR, e)
+        }
+    }
+
+    override suspend fun update(value: String) {
+        withContext(dispatcher) {
```
Comment (alex-bradbury-gds) [inline app/src/main/java/uk/gov/onelogin/appinfo/source/data/AppInfoLocalSourceImpl.kt:36]: not sure you need to do this?

Comment (BiancaMihaila) [reply app/src/main/java/uk/gov/onelogin/appinfo/source/data/AppInfoLocalSourceImpl.kt]: Took it out now

**Review by alex-bradbury-gds — APPROVED** [review]
Really great stuff 💪

**Pull request merged**